File: python/1845-seat-reservation-manager.py
import heapq

class SeatManager:

    def __init__(self, n: int):
        self.seats = [i for i in range(1, n + 1)]

    def reserve(self) -> int:
        return heapq.heappop(self.seats)

    def unreserve(self, seatNumber: int) -> None:
        heapq.heappush(self.seats, seatNumber)

File: javascript/1845-seat-reservation-manager.js

class SeatManager {
    /**
    * MinHeap
    * Time O(n*log(n)) | Space O(n)
    * https://leetcode.com/problems/seat-reservation-manager/
    * @param {number} n
    */
    constructor(n) {
        this.unreserved = new MinPriorityQueue({
            compare: (a, b) => a - b
        });

        for (let i = 1; i < n + 1; i++) {
            this.unreserved.enqueue(i);
        }
    }

    /**
     * Time O(log(n)) | Space O(1)
     * @return {number}
     */
    reserve() {
        const minAvailableSeat = this.unreserved.dequeue();
        return minAvailableSeat;
    }

    /** 
     * Time O(log(n)) | Space O(1)
     * @param {number} seatNumber
     * @return {void}
     */
    unreserve(seatNumber) {
        this.unreserved.enqueue(seatNumber);
    }
}
